python/lsst/afw/table/catalogMatches.py

```diff
@@ -254,3 +256,112 @@ def copyAliasMapWithPrefix(inSchema, outSchema, prefix=""):
         outSchema.getAliasMap().set(prefix + k, prefix + v)
 
     return outSchema
+
+
+def reindexCatalog(catalog, indices, deep=True):
+    """Apply a numpy index array to an afw Catalog
+
+    Parameters
+    ----------
+    catalog : subclass of `lsst.afw.table.BaseCatalog`
+        Catalog to reindex.
+    indices : `numpy.ndarray` of `int`
+        Index array.
+
+    Returns
+    -------
+    new : subclass of `lsst.afw.table.BaseCatalog`
+        Reindexed catalog. Records are shallow copies of those in ``catalog``.
+    """
+    new = type(catalog)(catalog.schema)
+    new.reserve(len(indices))
+    for ii in indices:
+        _rec = catalog[int(ii)]
+        new.extend([_rec], deep=deep)
+    return new
+
+
+def matchCatalogsExact(catalog1, catalog2, patch1=None, patch2=None):
+    """Match two catalogs derived from the same mergeDet catalog
+
+    When testing downstream features, like deblending methods/parameters
+    and measurement algorithms/parameters, it is useful to to compare
+    the same sources in two catalogs. In most cases this must be done
+    by matching on either RA/DEC or XY positions, which occassionally
+    will mismatch one source with another.
+
+    For a more robust solution, as long as the downstream catalog is
+    derived from the same mergeDet catalog, exact source matching
+    can be done via the unique ``(parent, deblend_peakID)``
+    combination. So this function performs this exact matching for
+    all sources both catalogs.
+
+    Parameters
+    ----------
+    catalog1, catalog2: subclass of `lsst.afw.table.BaseCatalog`
+        The two catalogs to merge
+
+    patch1, patch2: array of int
+        Patch for each row, converted into an integer.
+        In the gen3 butler this is done already, in gen2
+        it is recommended to use `patch2Int`, assuming that
+        the patches are the same structure as HSC, that range
+        from '0,0' to '9,9'.
+
+    Returns
+    -------
+    result: tuple of `lsst.afw.table.BaseCatalog`
+        Both catalogs reordered such that all rows are
+        matched to the same source.
+    """
+    # Only match the individual sources, the parents will
+    # already be matched by the mergeDet catalog
+    sidx1 = catalog1["parent"] != 0
+    sidx2 = catalog2["parent"] != 0
+
+    # Create the keys used to merge the catalogs
+    parents1 = np.array(catalog1["parent"][sidx1])
+    peaks1 = np.array(catalog1["deblend_peakId"][sidx1])
```

Comment:
You're using column names specific to certain processing, which I think is counter to the concept of afw as a low-level library. Can you make the column names to be function parameters? If not, I think this needs to get hoisted up to a more suitable package.

Reply:
I think that this particular function is general in the sense that this is the _only_ way to exactly match two source catalogs built from the same `mergeDet` catalog. So it seems to fit in this module to me. I _could_ make it more general and just allow the user to specify the keys to match on, but I'm not sure that would be useful.

Reply:
The point is that `mergeDet` catalogs are generated in pipe_tasks. afw shouldn't know about details from pipe_tasks or require a specific catalog schema: it is a general-purpose toolkit.

Reply:
Hmm.  The `mergeDet` concept is defined only far downstream in `pipe_tasks`, so I think I'm with Paul on this not really belonging in afw, but putting it in the same module or package as `MergeDetectionsTask` seems more appropriate than trying to invent something more general/hypothetical for this to claim to do.